### DISCUSSION TITLE
Issues trying to convert my AdventOfCode code to use Parsers

### DISCUSSION THREAD

**Discussion by oliverfoggin** [pointfreeco/swift-parsing#264] [Q&A] [answered]
Hey,

I'm very new with Parsers in this use case.

I've answered all the AoC days up to now using just raw Swift but wanted to convert what I've got to use Parsers.

Anyway, I'm trying to do day one which has an input like...

```
1000
2000
3000

4000

5000
6000

7000
8000
9000

10000
```

Copying the style that Luke Redpath did here... https://github.com/pointfreeco/swift-parsing/discussions/260

I tried to create a Parser like this...

```
private enum Parsers {
	static let calories = Parse({ $0 }) {
		Many {
			Int.parser()
		} separator: {
			Whitespace(1, .vertical)
		} terminator: {
			Whitespace(1, .vertical)
		}
	}
}
```

But I'm getting an Error on the line `Int.parser()` "Ambiguous use of 'parser(of:radix:)'".

If I change it to something like...

```
Many {
  Int.parser()
  ""
}
```

The error goes away.

Do I need to provide the `of:` and `radix:` parameters here? And if so... what are they?

Thanks

**Comment by tgrapperon** [comment, marked as answer]
Hey @oliverfoggin! Sometimes, the compiler can have difficulties picking the `Input` type. In your case, you can nudge it in the right direction using the `of:` argument of `Int.parser`:
```swift
private enum Parsers {
  static let calories = Parse({ $0 }) {
    Many {
      Int.parser(of: Substring.self)
    } separator: {
      Whitespace(1, .vertical)
    } terminator: {
      Whitespace(1, .vertical)
    }
  }
}
```
`SubString.self` is already the default of this parameter, but it is not enough in this configuration for the compiler apparently.
The radix defines the "base" that is used, and allows to parse the hexadecimal "5F" as "95" in base 10 for example. In your case, you can keep the default value of 10.

BTW, the `Parse ({ $0 })` argument has no effect here and you can remove the part in parenthesis if you want (or the whole `Parse` even and directly use `Many`).

**Reply by oliverfoggin** [reply]
Thanks! And yeah, in my actual code I'm using the transform to sum up the total of the ints from the `Many` array. 😄

**Reply by tgrapperon** [reply]
As an aside, the `transform` argument is there to allow converting the parsed tuple into another type. While it is possible to perform some business logic work here, I would probably only use it to project (or convert when parse-printing) the input into a workable data type, and then work with this data "out of line". In your case, the `[Int]` array is perfectly workable, and no conversion is required.

Again, it will work fine summing the values in the `transform` block, especially in this Advent of Code context, but I'm not sure this is semantically correct place to do this in a more "serious" project where you would want to separate parsing and business logic, for example for testability reasons.

**Comment by lihuizhao01887152** [comment]
你好，我已经收到你的邮件，我一定会仔细阅读，感谢你的来信，祝好！——by zlh

**Comment by Sanbingo** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
